### PR TITLE
HelixAdmin APIs and pipeline changes to support Helix Node Swap 

### PR DESCRIPTION
**Pull request by zpinto**
### Issues

- [x]  HelixAdmin.setInstanceOperation and addInstance sanity checks for SWAP_IN & SWAP_OUT #2662 

     - Can only set instance to SWAP_OUT
     - Can only add instance with the same logicaId as another if the already existing has InstanceOperation set to SWAP_OUT and there is not more than 1 already.

- [x]  Create HelixAdmin.isSwapReadyToComplete and completeSwapIfReady(returns if swap finished)  #2662 

    - completeSwapIfFinished will check if the swap is ready to complete. If ready, mark the SWAP_OUT instance as `HELIX_ENABLED: false` and clear the `INSTANCE_OPERATION` for SWAP_IN node.
  
- [x] Change WAGED assignment tie breaker logic to use logicalId when present #2662 

     - May want to make this configurable, as it will lead to shuffling when this change is deployed

- [x]  Refactor DelayedAutoRebalancerUtil and DelayedAutoRebalancer.computeBestPossibleStateForPartition
 to filter out SWAP_IN node because we will add SWAP_IN node to the end of preference list in computeBestPossibleStateForPartition until the swap is complete #2662 
     - Adding it to the end of the preference list will ensure that the SWAP_IN node only carries secondary states
     - When SWAP_IN node has identical Current State to SWAP_OUT, isSwapFinished completeSwapIfFinished will return true

### Description

Adding Helix Admin APIs and updating existing setInstanceOperation and addInstance implementations to support Helix Node Swap.

- setInstanceOperation: Add sanity check to disallow setting the `INSTANCE_OPERATION` to `SWAP_IN`, as this should be set when we addInstance to cluster. We can't have a case where where there are two instances in the cluster with the same logicalId unless there is exactly 1 with `SWAP_OUT` and 1 with `SWAP_IN` `INSTANCE_OPERATION`.
- addInstance: We now check to ensure that there is not already an instance in the cluster with the same logicalId. We only allow adding an instance with the same logicalId if the existing instance has `SWAP_OUT` `INSTANCE_OPERATION` or the add instance is being added in the disable state.
- canSwapComplete: Checks if the swap of two instances is ready to mark completed. It takes either the `SWAP_OUT` or `SWAP_IN` instance as an argument. It does the following checks:
  1. Both instances must be alive.
  2. Both instances must only have one session and not be carrying over from a previous session.
  3. Both instances must have no pending messages.
  4. Both instances cannot have partitions in the ERROR state
  5. SwapIn instance must have correct state for all partitions that are currently assigned to the SwapOut instance.
- completeSwapPossible: Checks to see if the swap is ready to complete using isSwapReadyToComplete. If ready, it will clear the SwapIn node's `INSTANCE_OPERATION` and disable the SwapOut node.

Add ability for up to 2 nodes with the same logicalId to be added to the cluster at the same time when a SWAP is happening.
During all paritionAssignment for WAGED and DelayedAutoRebalancer, we select just one instance for each logicalId.

Cases:
- If only one instance with logicalId, it passes through the filter
- If two instances with logicalId, SWAP_OUT passes through the filter if the other node is SWAP_IN. If SWAP_IN node passes through the filter when the SWAP is marked complete by setting the InstanceOperation of the SWAP_IN node to null.

When SWAP_OUT nodes are the instances which pass through the filter and there are corresponding SWAP_IN nodes which are live and enabled, we add them to the end of all preference lists which contain the SWAP_OUT node following computeIdealState. This ensures that they are in secondTopState for all StateModels that can only have a single replica in the topState. For stateModels that allow multiple replicas of the topState, the SWAP_IN node could host topState replicas before the SWAP is completed.

Once the swap is complete and SWAP_IN node has InstanceOperation set to null, SWAP_IN node becomes assignable and WAGED and DelayedAutoRebalancer + CRUSHED will computePartitionAssignment and generate IdealState with SWAP_IN instance in same index of the preferenceList as the SWAP_OUT instance used to be in. This means that not only will SWAP_IN node have all the same partitions that the SWAP_OUT node had, but also the same states.

### Tests

- [x] testAddingNodeWithSwapOutInstanceOperation
- [x] testAddingNodeWithSwapOutNodeInstanceOperationUnset
- [x] testNodeSwapWithNoSwapOutNode
- [x] testNodeSwapSwapInNodeNoInstanceOperationEnabled
- [x] testNodeSwapSwapInNodeWithAlreadySwappingPair
- [x] testNodeSwapWrongFaultZone
- [x] testNodeSwapWrongCapacity
- [x] testNodeSwapNoTopologySetup
- [x] testNodeSwap
- [x] testNodeSwapSwapInNodeNoInstanceOperationDisabled
- [x] testNodeSwapCancelSwapWhenReadyToComplete
- [x] testNodeSwapAfterEMM
- [x] testNodeSwapWithSwapOutInstanceDisabled
- [x] testNodeSwapAddSwapInFirstEnabledBeforeSwapOutSet(
- [x] testNodeSwapAddSwapInFirstEnableBeforeSwapOutSet
- [x] testUnsetInstanceOperationOnSwapInWhenAlreadyUnsetOnSwapOut
- [x] testNodeSwapAddSwapInFirst

```
➜  helix git:(helix_node_swap_apis) ✗ mvn test -o -Dtest=TestInstanceOperation -pl=helix-core
[INFO] Scanning for projects...
[WARNING] 
[WARNING] Some problems were encountered while building the effective model for org.apache.helix:metrics-common:bundle:1.3.2-SNAPSHOT
[WARNING] 'build.pluginManagement.plugins.plugin.(groupId:artifactId)' must be unique but found duplicate declaration of plugin org.apache.maven.plugins:maven-javadoc-plugin @ org.apache.helix:helix:1.3.2-SNAPSHOT, /Users/zapinto/Documents/git2/zpinto/helix/pom.xml, line 678, column 17
[WARNING] 
[WARNING] Some problems were encountered while building the effective model for org.apache.helix:helix:pom:1.3.2-SNAPSHOT
[WARNING] 'build.pluginManagement.plugins.plugin.(groupId:artifactId)' must be unique but found duplicate declaration of plugin org.apache.maven.plugins:maven-javadoc-plugin @ line 678, column 17
[WARNING] 
[WARNING] It is highly recommended to fix these problems because they threaten the stability of your build.
[WARNING] 
[WARNING] For this reason, future Maven versions might no longer support building such malformed projects.
[WARNING] 
[INFO] 
[INFO] --------------------< org.apache.helix:helix-core >---------------------
[INFO] Building Apache Helix :: Core 1.3.2-SNAPSHOT
[INFO]   from pom.xml
[INFO] -------------------------------[ bundle ]-------------------------------
[INFO] 
[INFO] --- enforcer:1.4.1:enforce (enforce-maven-version) @ helix-core ---
[INFO] 
[INFO] --- enforcer:1.4.1:enforce (enforce-java-version) @ helix-core ---
[INFO] 
[INFO] --- enforcer:1.4.1:enforce (enforce-output-timestamp-property) @ helix-core ---
[INFO] 
[INFO] --- jacoco:0.8.6:prepare-agent (default) @ helix-core ---
[INFO] argLine set to -javaagent:/Users/zapinto/.m2/repository/org/jacoco/org.jacoco.agent/0.8.6/org.jacoco.agent-0.8.6-runtime.jar=destfile=/Users/zapinto/Documents/git2/zpinto/helix/helix-core/target/jacoco.exec
[INFO] 
[INFO] --- remote-resources:1.7.0:process (process-resource-bundles) @ helix-core ---
[INFO] Preparing remote bundle org.apache:apache-jar-resource-bundle:1.4
[INFO] Copying 3 resources from 1 bundle.
[INFO] 
[INFO] --- resources:3.2.0:resources (default-resources) @ helix-core ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] Copying 4 resources
[INFO] Copying 0 resource
[INFO] Copying 3 resources
[INFO] The encoding used to copy filtered properties files have not been set. This means that the same encoding will be used to copy filtered properties files as when copying other filtered resources. This might not be what you want! Run your build with --debug to see which files might be affected. Read more at https://maven.apache.org/plugins/maven-resources-plugin/examples/filtering-properties-files.html
[INFO] 
[INFO] --- compiler:3.10.1:compile (default-compile) @ helix-core ---
[INFO] Nothing to compile - all classes are up to date
[INFO] 
[INFO] --- bundle:5.1.4:manifest (bundle-manifest) @ helix-core ---
[WARNING] Manifest org.apache.helix:helix-core:bundle:1.3.2-SNAPSHOT : Unused Import-Package instructions: [org.apache.logging.log4j*, org.apache.logging.slf4j*] 
[INFO] Writing manifest: /Users/zapinto/Documents/git2/zpinto/helix/helix-core/target/classes/META-INF/MANIFEST.MF
[INFO] 
[INFO] --- resources:3.2.0:testResources (default-testResources) @ helix-core ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Using 'UTF-8' encoding to copy filtered properties files.
[INFO] Copying 15 resources
[INFO] Copying 3 resources
[INFO] 
[INFO] --- compiler:3.10.1:testCompile (default-testCompile) @ helix-core ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 498 source files to /Users/zapinto/Documents/git2/zpinto/helix/helix-core/target/test-classes
[INFO] /Users/zapinto/Documents/git2/zpinto/helix/helix-core/src/test/java/org/apache/helix/integration/TestCarryOverBadCurState.java: Some input files use or override a deprecated API.
[INFO] /Users/zapinto/Documents/git2/zpinto/helix/helix-core/src/test/java/org/apache/helix/integration/TestCarryOverBadCurState.java: Recompile with -Xlint:deprecation for details.
[INFO] /Users/zapinto/Documents/git2/zpinto/helix/helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java: Some input files use unchecked or unsafe operations.
[INFO] /Users/zapinto/Documents/git2/zpinto/helix/helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java: Recompile with -Xlint:unchecked for details.
[INFO] 
[INFO] --- surefire:3.0.0-M3:test (default-test) @ helix-core ---
[INFO] 
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running org.apache.helix.integration.rebalancer.TestInstanceOperation
Start zookeeper at localhost:2183 in thread main

...

START TestInstanceOperation.testAddingNodeWithSwapOutInstanceOperation() at Wed Nov 01 16:07:33 PDT 2023
START TestInstanceOperation.testAddingNodeWithSwapOutNodeInstanceOperationUnset() at Wed Nov 01 16:07:35 PDT 2023
START TestInstanceOperation.testNodeSwapWithNoSwapOutNode() at Wed Nov 01 16:07:37 PDT 2023
START TestInstanceOperation.testNodeSwapSwapInNodeNoInstanceOperationEnabled() at Wed Nov 01 16:07:39 PDT 2023
START TestInstanceOperation.testNodeSwapSwapInNodeWithAlreadySwappingPair() at Wed Nov 01 16:07:43 PDT 2023
START TestInstanceOperation.testNodeSwapNoTopologySetup() at Wed Nov 01 16:07:47 PDT 2023
START TestInstanceOperation.testNodeSwapWrongFaultZone() at Wed Nov 01 16:07:49 PDT 2023
START TestInstanceOperation.testNodeSwapWrongCapacity() at Wed Nov 01 16:07:55 PDT 2023
START TestInstanceOperation.testNodeSwap() at Wed Nov 01 16:07:57 PDT 2023
START TestInstanceOperation.testNodeSwapSwapInNodeNoInstanceOperationDisabled() at Wed Nov 01 16:08:05 PDT 2023
START TestInstanceOperation.testNodeSwapCancelSwapWhenReadyToComplete() at Wed Nov 01 16:08:15 PDT 2023
START TestInstanceOperation.testNodeSwapAfterEMM() at Wed Nov 01 16:08:26 PDT 2023
START TestInstanceOperation.testNodeSwapWithSwapOutInstanceDisabled() at Wed Nov 01 16:08:36 PDT 2023
START TestInstanceOperation.testNodeSwapAddSwapInFirstEnabledBeforeSwapOutSet() at Wed Nov 01 16:08:47 PDT 2023
START TestInstanceOperation.testNodeSwapAddSwapInFirstEnableBeforeSwapOutSet() at Wed Nov 01 16:08:50 PDT 2023
START TestInstanceOperation.testUnsetInstanceOperationOnSwapInWhenAlreadyUnsetOnSwapOut() at Wed Nov 01 16:08:55 PDT 2023
START TestInstanceOperation.testNodeSwapAddSwapInFirst() at Wed Nov 01 16:08:59 PDT 2023
AfterClass: TestInstanceOperation called.
Shut down zookeeper at port 2183 in thread main
[INFO] Tests run: 24, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 238.321 s - in org.apache.helix.integration.rebalancer.TestInstanceOperation
[INFO] 
[INFO] Results:
[INFO] 
[INFO] Tests run: 24, Failures: 0, Errors: 0, Skipped: 0
[INFO] 
[INFO] 
[INFO] --- jacoco:0.8.6:report (generate-code-coverage-report) @ helix-core ---
[INFO] Loading execution data file /Users/zapinto/Documents/git2/zpinto/helix/helix-core/target/jacoco.exec
[INFO] Analyzed bundle 'Apache Helix :: Core' with 947 classes
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  03:36 min
[INFO] Finished at: 2023-10-30T12:32:08-04:00
[INFO] ------------------------------------------------------------------------
```

### Changes that Break Backward Compatibility (Optional)

- We change the compareTo method in AssignableNode to use logicalId to make sure partitionAssignment will be the same after switching to SWAP_IN node. This is only used for tie breaking, so impact should not be large on already existing WAGED cluster assignment; however, it can cause some replica shuffling.

### Commits

- My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Code Quality

- My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/main/java/org/apache/helix/HelixAdmin.java [hunk 1]
```diff
@@ -747,6 +756,26 @@ Map<String, Boolean> validateInstancesForWagedRebalance(String clusterName,
...
+  /**
+   * Check to see if swapping between two instances is ready to be completed and completed it if
+   * ready. Either the swapOut or swapIn instance can be passed in.
+   *
+   * @param clusterName  The cluster name
+   * @param instanceName The instance that is being swapped out or swapped in
+   * @return True if the swap is ready to be completed and was completed successfully, false
+   * otherwise.
+   */
+  boolean completeSwapIfReady(String clusterName, String instanceName);
```
Comment (MarkGaox) [inline helix-core/src/main/java/org/apache/helix/HelixAdmin.java:777]: Naming

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java [hunk 2]
```diff
@@ -410,13 +472,260 @@ public ZNRecord update(ZNRecord currentData) {
...
+    }
+
+    return null;
+  }
+
+  /**
+   * Check to see if swapping between two instances is ready to be completed. Checks: 1. Both
+   * instances must be alive. 2. Both instances must only have one session and not be carrying over
+   * from a previous session. 3. Both instances must have no pending messages. 4. Both instances
+   * cannot have partitions in the ERROR state 4. SwapIn instance must have correct state for all
```
Comment (MarkGaox) [inline helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java:525]: Typo: 5

File: helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java [hunk 3]
```diff
@@ -410,13 +472,260 @@ public ZNRecord update(ZNRecord currentData) {
...
+   * partitions that are currently assigned to the SwapOut instance.
+   * TODO: We may want to make this a public API in the future.
```
Comment (xyuanlu) [inline helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java:527]: Add check ideal state and CS as well. Possible that when no pending msg, there will be partition assigned in the future.

Comment (zpinto) [reply helix-core/src/main/java/org/apache/helix/manager/zk/ZKHelixAdmin.java:527]: I don't think this will work in every case because the ideal state is not persisted in ZK without PERSIST_BEST_POSSIBLE_STATE set to true. I think that only checking current state and no pending messages is sufficient because if the replica from the ideal state is not already being bootstrapped on SWAP_OUT node then it will be sent directly to SWAP_IN node right after the swap completes.

If load balance is causing the new replica to be assigned to SWAP_OUT node then if the ST message never gets to the SWAP_OUT node before completing the SWAP it is okay because once we switch to the SWAP_IN node we have not dropped the replica from its old location yet.

This means that it will still be N -> N+1 -> N for the newly assigned replica.

**Review by xyuanlu — APPROVED** [review]

**Review by desaikomal — COMMENTED** [review]
these are just my initial comments, i will do next pas soon..

File: helix-core/src/main/java/org/apache/helix/HelixAdmin.java [hunk 4]
```diff
@@ -747,6 +756,26 @@ Map<String, Boolean> validateInstancesForWagedRebalance(String clusterName,
    */
   boolean isEvacuateFinished(String clusterName, String instancesNames);
 
+  /**
+   * Check to see if swapping between two instances can be completed. Either the swapOut or
+   * swapIn instance can be passed in.
+   * @param clusterName  The cluster name
+   * @param instanceName The instance that is being swapped out or swapped in
+   * @return True if the swap is ready to be completed, false otherwise.
```
[the remark this answers is not in the document]
Comment (zpinto) [reply helix-core/src/main/java/org/apache/helix/HelixAdmin.java:764]: Instance name in this case will be one of the unique instance names(same as InstanceConfig znode ID) involved in the swap. Either SWAP_OUT or SWAP_IN. This API is intended to be used with the PerInstanceAccessor in helix-rest. For that endpoint you pass one instanceName.

Because we can find the matching swap instance for either SWAP_OUT or SWAP_IN node, we can take either as the instance name.

"swap" is referring to the operation between two instances which can be deduced from providing one of the instances involved in the swap.

File: helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java [hunk 5]
```diff
@@ -858,6 +886,42 @@ private void updateDisabledInstances(Collection<InstanceConfig> instanceConfigs,
     }
   }
 
+  private void updateSwappingInstances(Collection<InstanceConfig> instanceConfigs,
```
[the remark this answers is not in the document]
Comment (zpinto) [reply helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java:889]: `updateSwappingInstances` is called by `doRefresh` which is a `synchronized` method; therefore, only one thread will call this method.

All of the other places this method is called from are in this classes setters which are only used in unit tests and HelixUtil code which is not invoked in the controller.

We could technically make this method `synchronized` as well just to be safe. Initially I didn't do this, as I was following standard for `updateDisabledInstances`, etc.

[the remark this answers is not in the document]
Comment (desaikomal) [reply helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java:889]: another quick question: how do you make sure that incoming host has similar capacity as outgoing and not sure about FaultDomain as well

[the remark this answers is not in the document]
Comment (zpinto) [reply helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java:889]: The assumption is that the if the instance is added with the same logicalId as an existing instance, it should meet the criteria of having the same zone and capacity. We ca definitely add a check to fail early if trying to `addInstance` without adhering to the swap requirements.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/PerInstanceAccessor.java [hunk 6]
```diff
@@ -435,6 +435,11 @@ public Response updateInstance(@PathParam("clusterId") String clusterId,
       case setInstanceOperation:
          admin.setInstanceOperation(clusterId, instanceName, state);
          break;
+        case completeSwapIfPossible:
```
[the remark this answers is not in the document]
Comment (zpinto) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/PerInstanceAccessor.java:438]: This apparently is the correct indentation according to helix style configuration; however, the others are incorrect. I didn't change the others because want to keep the diff only on my changes.

File: helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (zpinto) [reply helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java:889]: @desaikomal , I added a check to not let SWAP_IN node join if it is not in the same FAULT_ZONE or does not have the same capacity.

File: helix-core/src/main/java/org/apache/helix/HelixAdmin.java [hunk 7]
```diff
@@ -747,6 +756,26 @@ Map<String, Boolean> validateInstancesForWagedRebalance(String clusterName,
...
+   */
+  boolean canSwapBeCompleted(String clusterName, String instanceName);
```
[the remark this answers is not in the document]
Comment (zpinto) [reply helix-core/src/main/java/org/apache/helix/HelixAdmin.java]: changed, thanks for the suggestion

File: helix-core/src/main/java/org/apache/helix/controller/rebalancer/DelayedAutoRebalancer.java [hunk 8]
```diff
@@ -113,9 +114,16 @@ public IdealState computeNewIdealState(String resourceName,
       allNodes = clusterData.getAllInstances();
     }
 
+    Set<String> allNodesDeduped = DelayedRebalanceUtil.filterOutInstancesWithDuplicateLogicalIds(
```
[the remark this answers is not in the document]
Comment (zpinto) [reply helix-core/src/main/java/org/apache/helix/controller/rebalancer/DelayedAutoRebalancer.java:117]: Added a TODO for this. We can do it when we refactor to use states instead of InstanceOperation.

File: helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java [hunk 9]
```diff
@@ -617,6 +625,24 @@ public Set<String> getDisabledInstances() {
...
+  public Map<String, String> getSwapOutToSwapInInstancePairs() {
+    return Collections.unmodifiableMap(_swapOutInstanceNameToSwapInInstanceName);
+  }
+
+  /**
+   * Get all the enabled and live SWAP_IN instances.
+   *
+   * @return a set of SWAP_IN instanceNames that have a corresponding SWAP_OUT instance.
+   */
+  public Set<String> getEnabledLiveSwapInInstanceNames() {
```
[the remark this answers is not in the document]
Comment (zpinto) [reply helix-core/src/main/java/org/apache/helix/controller/dataproviders/BaseControllerDataProvider.java:642]: Discussed offline. 

TLDR: Sanity checks ensure that we can't have a case where there are two instances with the same logicalId and both have InstanceOperation unset. This allows for setting either SWAP_OUT node or SWAP_IN node instance operation first.

File: helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/PerInstanceAccessor.java (hunk 6 above)
Comment (xyuanlu) [inline helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/PerInstanceAccessor.java:438]: nit: alignment?

Comment (xyuanlu) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/PerInstanceAccessor.java:438]: Also do we have canCompleteSwap?

Comment (zpinto) [reply helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/PerInstanceAccessor.java:443]: Added canCompleteSwap and fixed formatting.

**Review by xyuanlu — APPROVED** [review]

**Comment by zpinto** [comment]
CI failure is due to recurring flaky test testDistributedController https://github.com/apache/helix/issues/2498

This PR is ready to be merged.

Final Commit Message:

Add ability for up to 2 nodes with the same logicalId to be added to the cluster at the same time when a SWAP is happening.
During all paritionAssignment for WAGED and DelayedAutoRebalancer, we select just one instance for each logicalId. Achieves n -> n+1 for all replicas on SWAP_OUT node and back n when SWAP is marked complete, making it cancelable.

Adding and updating Helix Admin APIs to support swap operation:
- setInstanceOperation
- addInstance
- canCompleteSwap
- completeSwapIfPossible
